collectors/google_pubsub/pubsub_collector.go

```diff
@@ -36,19 +60,79 @@ func (collector *PubSubCollector) HandleMessage(message *pubsub.Message) error {
 	return result.Err
 }
 
+func (collector *PubSubCollector) PollForMessages() {
+	// QUESTION - should I add the channel usage back in here like in AWS Collector?
+	for {
+		if !collector.BaseCollector.Sleeping {
+			fmt.Println("Processing messages...")
+			messages, err := collector.GetMessages()
+			if err != nil {
+				fmt.Println("Error recieving messages - retrying")
+				collector.ProcessExponentialBackoff(err)
+			}
+			for _, msg := range messages {
+				collector.BusinessProcessor(msg.MessageBody)
+			}
+			fmt.Printf("Recieved messages - %s\n", messages)
+		}
+	}
+}
+
 func (collector *PubSubCollector) GetMessages() ([]collectors.MessageWrapper, error) {
-	var err error
+	// There's a built in deadline on the subscription that will orphan the message in the subscription.
+	subscription := collector.PubSubClient.Subscription(collector.SourceSubscription)
 	messages := make([]collectors.MessageWrapper, 0)
 
+	// NOTE - HTF is this referencing the subscription.Mutex?!?!?!?!
+	var mu sync.Mutex
+	err := subscription.Receive(collector.Context, func(ctx context.Context, msg *pubsub.Message) {
+		mu.Lock()
+		defer mu.Unlock()
+
+		received := 0
+		fmt.Printf("Got message: %q\n", string(msg.Data))
+		messages = append(messages, collectors.MessageWrapper{
+			MessageBody: msg.Data,
+			Retry:       false,
+			Fatal:       false,
+		})
+		msg.Ack()
+
+		received++
+		if received >= collector.PollingLimit {
+			collector.Cancel()
+		}
+	})
+	if err != nil {
+		fmt.Printf("Could not recieve messages from - %s : %v", collector.SourceSubscription, err)
+	}
 	return messages, err
 }
 
 func (collector *PubSubCollector) PublishMessage(message *collectors.MessageWrapper, delaySeconds int64, errFlag bool) error {
-	var err error
-	return err
+	topic := collector.PubSubClient.Topic(collector.SourceTopic)
+	if errFlag {
+		topic = collector.PubSubClient.Topic(collector.ErrorTopic)
+	}
+	// Clean up goroutines for batching and sending messages
+	defer topic.Stop()
+
+	result := topic.Publish(collector.Context, &pubsub.Message{
+		Data: message.MessageBody,
+	})
+	// This will block until a result is returned from Google
+	id, err := result.Get(collector.Context)
+	if err != nil {
+		// TODO - need some way to retry publishes
```

Comment:
For now maybe we should log it and try to come up with a better way later, but eveb if we retry we cannot retry forever...

```suggestion
		// TODO - need some way to retry publishes
		log.Printf("Error publishing retry message: %+v\n", pubsub.Message)
```